File: src/main/java/io/github/bluesheep2804/selenechat/listener/ChatListenerSpigot.java
package io.github.bluesheep2804.selenechat.listener;

import com.google.common.io.ByteArrayDataOutput;
import com.google.common.io.ByteStreams;
import io.github.bluesheep2804.selenechat.SeleneChatSpigot;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;


public class ChatListenerSpigot implements Listener {
    @EventHandler
    public void onPlayerChat(AsyncPlayerChatEvent event) {
        String message = event.getMessage();
        Component returnMessage = ChatProcess.message(message);

        event.setMessage(LegacyComponentSerializer.legacySection().serialize(returnMessage));
        ByteArrayDataOutput out = ByteStreams.newDataOutput();
        out.writeUTF(event.getPlayer().getDisplayName());
        out.writeUTF(LegacyComponentSerializer.legacySection().serialize(returnMessage));
        SeleneChatSpigot.getInstance().sendPluginMessage(out.toByteArray());
    }
}

File: src/main/kotlin/io/github/bluesheep2804/selenechat/listener/ChatListenerSpigot.kt
package io.github.bluesheep2804.selenechat.listener

import io.github.bluesheep2804.selenechat.SeleneChat.config
import io.github.bluesheep2804.selenechat.SeleneChatSpigot
import io.github.bluesheep2804.selenechat.message.PluginMessage
import io.github.bluesheep2804.selenechat.player.SeleneChatPlayerSpigot
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent

class ChatListenerSpigot(private val plugin: SeleneChatSpigot) : Listener {
    @EventHandler
    fun onPlayerChat(event: AsyncPlayerChatEvent) {
        val message = event.message
        val sender = SeleneChatPlayerSpigot(event.player)

        val result = ChatListener.chat(message, sender)
        if (result is Component)
            event.message = LegacyComponentSerializer.legacySection().serialize(result)
        else
            event.isCancelled = true

        if (config.shouldSendPluginMessage) {
            plugin.sendPluginMessage(PluginMessage(event.message, sender).build())
        }
    }
}